### DISCUSSION TITLE
Could you please share a valid network config

### DISCUSSION THREAD

**Discussion by peng6662001** [cloud-hypervisor/cloud-hypervisor#5084] [Q&A] [answered]
Hi,cloud-hypervisor team,could you please share a valid network config?
This is my command:
sudo ./target/release/cloud-hypervisor \
        --kernel ../linux-cloud-hypervisor/arch/arm64/boot/Image  \
        --disk path=$ROOTFS path=/tmp/ubuntu-cloudinit.img  \
        --cmdline "console=hvc0 root=/dev/vda1 rw" \
        --cpus boot=4   \
        --memory size=0,shared=on  \
        --memory-zone id=mem0,size=1G,shared=on,host_numa_node=0 \
        --net "tap=,mac=,ip=,mask="
This is the result of ifconfig in cloud-hypervisor:
enp0s4: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        ether 2e:f6:81:84:bd:ac  txqueuelen 1000  (Ethernet)
        RX packets 7  bytes 586 (586.0 B)
        RX errors 0  dropped 7  overruns 0  frame 0
        TX packets 3  bytes 969 (969.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
Then I reference https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/macvtap-bridge.md
My tap device is /dev/tap16
If I run command with --net fd=3,mac=$mac 3<>$"$tapdevice",it return error:
Error booting VM: VmBoot(DeviceManager(CreateVirtioNet(TapError(IoctlError([2147767506](tel:2147767506), Os { code: 25, kind: Uncategorized, message: "Inappropriate ioctl for device" })))))
I can't understand 3<>$"$tapdevice",is it a right syntax?
Which document shall I reference more?
I want to browse internet in cloud-hypervisor

**Comment by sboeuf** [comment, marked as answer]
Hi, the correct reference document is https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/macvtap-bridge.md as you mentioned.
It's the simplest way to get external connectivity from your Cloud Hypervisor VM.
And yes, `--net fd=3,mac=$mac 3<>$"$tapdevice"` is the correct syntax if you're using `/bin/bash` as your shell.

**Reply by peng6662001** [reply]
@sboeuf @michael2012z 
Hi sboeuf and michael,I reference macvtap-bridge.md and write a test.sh,but it still returns error.
./test.sh 
95: macvtap0@enP9p3s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP mode DEFAULT group default qlen 500
    link/ether c2:67:4f:53:29:cb brd ff:ff:ff:ff:ff:ff
Error booting VM: VmBoot(DeviceManager(CreateVirtioNet(TapError(IoctlError(2147767506, Os { code: 25, kind: Uncategorized, message: "Inappropriate ioctl for device" })))))

Could you please check the test.sh:
#!/bin/bash
mac="c2:67:4f:53:29:cb"
host_net="enP9p3s0"

sudo ip link delete macvtap0
sudo ip link add link "$host_net" name macvtap0 type macvtap
sudo ip link set macvtap0 address "$mac" up
sudo ip link show macvtap0

tapindex=$(< /sys/class/net/macvtap0/ifindex)
tapdevice="/dev/tap$tapindex"

sudo chown "$UID.$UID" "$tapdevice"

sudo target/release/cloud-hypervisor \
        --kernel ../linux-cloud-hypervisor/arch/arm64/boot/Image \
        --disk path=../debian-sid-nocloud-arm64-daily.raw \
        --cpus boot=1 --memory size=512M \
        --cmdline "root=/dev/vda1 console=hvc0" \
        --net fd=3,mac=$mac 3<>$"$tapdevice"

**Reply by sboeuf** [reply]
Please run it with `strace -ff` so that we can figure out which ioctl is failing.

**Reply by rbradford** [reply]
@sboeuf The ioctl code is already embedded in the error: 2147767506 aka 0x800454D2 aka `TUNGETIFF`. This is the same as https://github.com/cloud-hypervisor/cloud-hypervisor/issues/4917 where the user gave up using macvtap in the end and just attached a plain tap.

**Reply by sboeuf** [reply]
Ah thanks for the pointer!
@peng6662001 something's wrong with your macvtap link. Please take a look at #4917 to use a bridge instead of macvtap.

**Reply by peng6662001** [reply]
I have referenced #4917,but get different results on x86 server and arm server,the VM's network can work on x86 server now,but I can't find ens5 on vm of arm64 server.
```
On x86 server:
  sudo ./ch-remote --api-socket=/tmp/ch-socket3 add-net tap=chtap1
  {"id":"_net2","bdf":"0000:00:05.0"}
  
  ll /tmp/ch-socket3 
  srwx------ 1 root root 0  2月  1 17:58 /tmp/ch-socket3=  


On  VM:
cloud@cloud:~$ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: ens5: <BROADCAST,MULTICAST> mtu 1500 qdisc noop state DOWN group default qlen 1000
    link/ether 2e:5a:37:8c:8e:0c brd ff:ff:ff:ff:ff:ff
    altname enp0s5
```
   
    
```
On arm64 server:
  sudo ./ch-remote --api-socket /tmp/ch-socket3 add-net tap=chtap1
  {"id":"_net2","bdf":"0000:00:05.0"}
  
  ll /tmp/ch-socket3 
  srwx------ 1 root root 0 Feb  1 09:45 /tmp/ch-socket3=


On VM:
cloud@cloud:~$ ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
```

**Reply by michael2012z** [reply]
Hi, @peng6662001 , in the guest kernel log (`dmesg`), is there anything wrong about the NIC driver?

**Reply by peng6662001** [reply]
@michael2012z There is no new kernel log about network on VM after I run add-net on host.

**Reply by michael2012z** [reply]
Hi, @peng6662001 

I am very sorry for my late response to this discussion.

Now I think I know the reason of the problem:
- Hotplug on AArch64 requires the usage of ACPI. 
- While ACPI requires UEFI booting. That means you must boot the VM from a firmware, not from the direct kernel. This restriction is only for AArch64. On X86, the VM booting from direct kernel can also play ACPI (and hotplug of course).
- You had `--kernel ../linux-cloud-hypervisor/arch/arm64/boot/Image` in your command line, right? So you booted the VM from a direct kernel, that's why the subsequent hotplug didn't work.

My suggestions:
- You can follow this guide page to build a EDK2 firmware on your arm server: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/docs/uefi.md . Replace the kernel image with the firmware binary, boot your VM and continue the test.
- You can also run an integration test case by playing `./scripts/dev_cli.sh tests --integration -- --test-filter test_macvtap_hotplug`. This way the CI script will build the EDK2 automatically in a docker container (the binary will be available at ~/workloads/CLOUDHV_EFI.fd, you can use it for your manually test) and run the macvtap hotplug test case.
- Check the code of the test case for more detail if necessary: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/d19a5d535607b6fe2aa712be0717bc927aad6de0/tests/integration.rs#L6296

**Reply by peng6662001** [reply]
@michael2012z Hi,I have tested test_macvtap_hotplug and research it.
I found the case uses ch-remote to add network.You can find details in [Macvtap-bridge-for-cloud-hypvervisor](https://github.com/AmpereComputing/ampere-lts-kernel/wiki/Macvtap-bridge-for-cloud-hypvervisor)
But I want to know how to use 3<>$"$tapdevice" in shell.
I have tested the script on x86 and aarch64 platform with kernel 5.15 and 6.2.0
Do you have another suggestion?

**Reply by adamliyi** [reply]
@peng6662001 , pleaser refer to https://github.com/cloud-hypervisor/cloud-hypervisor/issues/4917#issuecomment-1442834740.
To make '3<>/dev/tapX' work, we cannot run cloud hypervisor as sudo.

**Reply by peng6662001** [reply]
Thank you,I have verified it.
